### PR TITLE
feat: allow different admin repo by env variable

### PR DESCRIPTION
**Pull request by quike**
Issue: [230](https://github.com/github/safe-settings/issues/230)

Extract configuration to an external library so it can be updated via env variables. This should allow to select a different default admin organization.

This change permits to pass an environment variable for `ADMIN_REPO`, `SETTINGS_FILE_PATH` and `DEPLOYMENT_CONFIG_FILE` so they can be modified, like for example selecting as `ADMIN_REPO` the repository `.github` as other probot apps.

| Environment Variable               | Internal Reference       | Default Value             |
|------------------------------------|--------------------------|---------------------------|
| `ADMIN_REPO`         | `ADMIN_REPO`             | `admin`                   |
| `SETTINGS_FILE_PATH` | `SETTINGS_FILE_PATH`     | `settings.yml`            |
| `DEPLOYMENT_CONFIG_FILE`           | `DEPLOYMENT_CONFIG_FILE` | `deployment-settings.yml` |

Example:

```
ADMIN_REPO=global-config
SETTINGS_FILE_PATH=safe-settings.yml
DEPLOYMENT_CONFIG_FILE=safe-settings-deployment.yml
```

And the code will look within your `<ORG>/global-config/.github/safe-settings.yml` instead of default `<ORG>/admin/.github/settings.yml`

### CODE REVIEW

**Comment by quike** [comment]
@decyjphr  could you take a look to this? thanks.

**Comment by decyjphr** [comment]
Awesome, I'll try to get this merged today.

File: test/unit/lib/env.test.js
Comment (martinm82) [inline test/unit/lib/env.test.js]: Does it make sense to allow the "admin" repo be in a different organization? Maybe we should use here `GH_ORG` instead?

Comment (quike) [reply test/unit/lib/env.test.js]: What is `GH_ORG`? is that an already defined env variable? what do you mean with "to allow the "admin" repo be in a different organization", this change does not cover that only the customizable org/repo base. Is that a question for an enhancement? Thanks!

**Comment by quike** [comment]
@decyjphr could you approve running workflows to check if pipeline passes? Thanks a million.

Screen Shot 2022-12-13 at 11 53 15 AM

File: test/unit/lib/env.test.js
Comment (martinm82) [reply test/unit/lib/env.test.js]: I might misunderstand something but from what I know safe-settings works like that:
* create a GH app within `orgA`
* configure safe-settings probot app and set GH_ORG, APP_ID, WEBHOOK_SECRET, etc.

By doing that safe-settings will listen on events on GH_ORG (where the app was installed). If now you decide to set `ADMIN_ORG` on a different organization than `orgA` I guess certain things won't work, like getting events for repo `ADMIN_ORG:admin` from the GitHub App.

So to me it seems `ADMIN_ORG` should point always to `GH_ORG` or do I misunderstand something here?

Comment (quike) [reply test/unit/lib/env.test.js]: That GH_ORG is a *manifest* variable to set the app registered to an organization (like kind of owner/developer of the app) but it doesn't need to be set, as if not it gets sets to the user that requests it. It says who owns the app.

https://probot.github.io/docs/configuration/

```
The organization where you want to register the app in the app creation manifest flow. 
If set, the app is registered for an organization (https://github.com/organizations/ORGANIZATION/settings/apps/new), if not set, the GitHub app would be registered for the user account (https://github.com/settings/apps/new).
```

I think that is different than the env variable we are trying to add here, that is *behavioural*. Here we are trying to say: Instead of looking for an org or repo called `admin` we allow GitHub App owners (the GH_ORG or the user owner) to say that this instance of the app should look for a different org or repo as admin)

As an example I would say that several probot apps use `.github` repo as default kind of admin, where a profile file can be added so it is shown as a introduction when the user is in the org page. Based on that a bunch of apps are using that `.github` repo as admin/default, but safe-settings is looking for `admin` repo for such purpose. This change allows to set `.github` repo as admin, so we can have all in one place.

**Comment by decyjphr** [comment]
@quike Approved the actions workflow 👍🏾

**Comment by quike** [comment]
Thanks @decyjphr could you do it again and review the PR? sorry I added a change to set the rest of variables as env var as well.

File: test/unit/lib/env.test.js
Comment (martinm82) [reply test/unit/lib/env.test.js]: Sorry for the late response. 

To be clear: the authentication between the Probot app and the "`admin`" repo is achieved through the `APP_ID` and `PRIVATE_KEY` which is registered on `GH_ORG`, right?

If now `GH_ORG` != `ADMIN_ORG`, how will the Probot app be able to authenticate with the `ADMIN_ORG` (especially in an organization with private repos)?

Just that it is clear: I really love this change as one can finally use a different repository instead `admin` and especially the `CONFIG_PATH` which would allow us to move out the configuration to a different folder. I definitely will use this new feature once it is merged. I just do not understand yet the need for a new `ADMIN_ORG` variable as I believe it would be better to not allow different values here.

**Review by martinm82 — APPROVED** [review]

**Comment by quike** [comment]
Happy New Year! @decyjphr any chance you could take a look to this? thanks a million.

File: lib/configManager.js
Comment (martinm82) [inline lib/configManager.js:53]: Have you tested this change with a different path outside of `.github` folder? As described in the probot docs https://probot.github.io/api/latest/classes/context.Context.html#config the configuration needs to live inside `.github` folder:

> Name of the YAML file in the .github directory

See as well https://probot.github.io/docs/best-practices/#store-configuration-in-the-repository

Comment (quike) [reply lib/configManager.js:53]: @martinm82 Yes it worked. 

```
SAFE_SETTINGS_ADMIN_REPO=.github
SAFE_SETTINGS_CONFIG_PATH=.config/
SAFE_SETTINGS_SETTINGS_FILE_PATH=safe-settings.yml
```

test

```
INFO (server): POST http://localhost:3000/ - 202
INFO (event): Updating repo with topics github,probot,new-topic,another-topic,topic-12
    id: "edd0c130-a36c-11ed-9add-603722a602c7"
INFO (event): Updating repo with topics github,probot,new-topic,another-topic,topic-12
    id: "edd0c130-a36c-11ed-9add-603722a602c7"
INFO (event): Updating repo with topics github,probot,new-topic,another-topic,topic-12
    id: "ece7c0c0-a36c-11ed-86f0-6d726ae5924e"
INFO (event): Updating repo with topics github,probot,new-topic,another-topic,topic-12
    id: "ece7c0c0-a36c-11ed-86f0-6d726ae5924e"
INFO (event): Updating repo with topics github,probot,new-topic,another-topic,topic-12
    id: "14336a30-a36d-11ed-955b-e30a62367fc8"
INFO (event): Updating repo with topics github,probot,new-topic,another-topic,topic-12
    id: "14336a30-a36d-11ed-955b-e30a62367fc8"
INFO (http): POST / 200 - 13ms
INFO (http): POST / 200 - 8ms
INFO (server): POST http://localhost:3000/ - 200
INFO (event): Finding teams for empirl-sandbox/test
    id: "edd0c130-a36c-11ed-9add-603722a602c7"
INFO (http): POST / 200 - 9ms
INFO (http): POST / 200 - 7ms
INFO (server): POST http://localhost:3000/ - 200
INFO (event): Getting team with the parms {"name":"bots","permission":"push"}
    id: "edd0c130-a36c-11ed-9add-603722a602c7"
INFO (event): Getting team with the parms {"name":"code-owners","permission":"admin"}
    id: "edd0c130-a36c-11ed-9add-603722a602c7"
INFO (event): Repo test Passed Validation for pattern [a-zA-Z0-9_-]+
    id: "edd0c130-a36c-11ed-9add-603722a602c7"
INFO (event): adding team bots to repo test
    id: "edd0c130-a36c-11ed-9add-603722a602c7"
INFO (event): Finding teams for empirl-sandbox/test
    id: "edd0c130-a36c-11ed-9add-603722a602c7"
INFO (event): adding team code-owners to repo test
    id: "edd0c130-a36c-11ed-9add-603722a602c7"
INFO (event): Getting team with the parms {"name":"bots","permission":"push"}
    id: "edd0c130-a36c-11ed-9add-603722a602c7"
INFO (event): Getting team with the parms {"name":"code-owners","permission":"admin"}
    id: "edd0c130-a36c-11ed-9add-603722a602c7"
INFO (event): Repo test Passed Validation for pattern [a-zA-Z0-9_-]+
    id: "edd0c130-a36c-11ed-9add-603722a602c7"
INFO (event): adding team code-owners to repo test
    id: "edd0c130-a36c-11ed-9add-603722a602c7"
INFO (event): adding team bots to repo test
    id: "edd0c130-a36c-11ed-9add-603722a602c7"
INFO (event): Finding teams for empirl-sandbox/test
    id: "ece7c0c0-a36c-11ed-86f0-6d726ae5924e"
INFO (event): Getting team with the parms {"name":"bots","permission":"push"}
    id: "ece7c0c0-a36c-11ed-86f0-6d726ae5924e"
INFO (event): Getting team with the parms {"name":"code-owners","permission":"admin"}
    id: "ece7c0c0-a36c-11ed-86f0-6d726ae5924e"
INFO (event): Repo test Passed Validation for pattern [a-zA-Z0-9_-]+
    id: "ece7c0c0-a36c-11ed-86f0-6d726ae5924e"
...
```

**Comment by decyjphr** [comment]
@quike sorry for the delay. Will try to get this merged in a day. @martinm82 and @quike appreciate your effort and comments.

**Comment by quike** [comment]
@decyjphr @martinm82 
I followed @martinm82 's comments and I removed the addition of `ADMIN_ORG` as it was unnecessary for my original intentions. I removed as well the change of CONFIG_PATH (probot predefined `.github`) I looks like it is hardcoded in several places and I don't have the time to be debugging all those specific use cases (like the suborgs and unique repo configs). Hopefully someone with a use case can work on that to extract all those hardcoded strings into a more globalized reusable approach to avoid those duplication

So far I just added my original intentions: have a different base repo than `admin`. I also added (although I doubt no one will ever use) the ability to rename the config and deployment files. In my case I want to use safe-settings with same repo than other github apps configurations, and  a `settings.yml` or `deployment-settings.yml` within the `.github` folder could collide with other apps, so now I could rename them both via variable to `safe-settings.yml` and `safe-settings-deployment.yml` 😄 

Feel free to tell me if you need anything else.

File: test/unit/lib/env.test.js
Comment (quike) [reply test/unit/lib/env.test.js]: Removed ADMIN_ORG change

File: lib/env.js
Comment (decyjphr) [inline lib/env.js]: @quike Could we simplify the names of the env variables`SAFE_SETTINGS_ADMIN_REPO` and `SAFE_SETTINGS_SETTINGS_FILE_PATH` to `ADMIN_REPO` and `SETTINGS_FILE_PATH`?

**Review by decyjphr — DISMISSED** [review]

File: lib/env.js
Comment (quike) [reply lib/env.js]: @decyjphr sure, I added in that way as they are global env variables to avoid clash with any other existing variable in the system keeping the scope for safe settings only as for example `ADMIN_REPO` is quite generic and could be used by other app or component within a defined system. But let me change it if you prefer it that way.

Comment (quike) [reply lib/env.js]: @decyjphr @martinm82 done, can you guys review whenever you can? Thanks

Comment (decyjphr) [reply lib/env.js]: Looks good to me.

**Review by decyjphr — APPROVED** [review]

**Review by martinm82 — APPROVED** [review]
Looks great.

**Pull request merged**
